Source/WebKit/webpushd/WebClipCache.mm

```diff
@@ -0,0 +1,253 @@
+        if (!path)
+            return false;
+        return [[NSFileManager defaultManager] fileExistsAtPath:path];
+    }
+}
+
+static String webClipIdentifierForOrigin(const String& bundleIdentifier, const WebCore::SecurityOriginData& origin)
+{
+    RetainPtr<NSString> oldestMatchingIdentifier;
+    WallTime oldestMatchingCreationTime = WallTime::infinity();
+
+    @autoreleasepool {
+        NSArray *webClips = [UIWebClip webClips];
+
+        for (UIWebClip *webClip in webClips) {
+            NSString *identifier = webClip.identifier;
+            if (!identifier)
+                continue;
+
+            auto clipOrigin = WebCore::SecurityOriginData::fromURL(webClip.pageURL);
+            if (origin != clipOrigin)
+                continue;
+
+            if (bundleIdentifier != "com.apple.SafariViewService"_s && [webClip respondsToSelector:@selector(trustedClientBundleIdentifiers)]) {
+                if (![webClip.trustedClientBundleIdentifiers containsObject:(NSString *)bundleIdentifier])
+                    continue;
+            }
+
+            NSString *path = [UIWebClip pathForWebClipWithIdentifier:identifier];
+            if (!path)
+                continue;
+
+            auto maybeCreationTime = FileSystem::fileCreationTime(path);
+            if (!maybeCreationTime)
+                continue;
+            auto creationTime = *maybeCreationTime;
+
+            if (creationTime < oldestMatchingCreationTime) {
+                oldestMatchingIdentifier = identifier;
+                oldestMatchingCreationTime = creationTime;
+            }
+        }
+    }
+
+    return oldestMatchingIdentifier.get();
+}
+
+String WebClipCache::preferredWebClipIdentifier(const String& bundleIdentifier, const WebCore::SecurityOriginData& origin)
+{
+    bool dirty = false;
+
+    auto it = m_preferredWebClipIdentifiers.find(std::make_tuple(bundleIdentifier, origin));
+    if (it != m_preferredWebClipIdentifiers.end()) {
+        auto identifier = it->value;
+        if (webClipExists(identifier))
+            return identifier;
+
+        m_preferredWebClipIdentifiers.remove(it);
+        dirty = true;
+    }
+
+    auto webClipIdentifier = webClipIdentifierForOrigin(bundleIdentifier, origin);
+    if (!webClipIdentifier.isEmpty()) {
+        m_preferredWebClipIdentifiers.set(std::make_tuple(bundleIdentifier, origin), webClipIdentifier);
+        dirty = true;
+    }
+
+    if (dirty)
+        persist();
+
+    return webClipIdentifier;
+}
+
+// Loads path as a plist and returns it if it has the shape [["a", "b", "c"], ["d", "e", "f"], ...]
+static RetainPtr<NSArray> loadWebClipCachePropertyList(NSString *path)
+{
+    @autoreleasepool {
+        NSError *error = nil;
+        NSData *data = [NSData dataWithContentsOfFile:path options:NSDataReadingMappedIfSafe error:&error];
+        if (!data) {
+            RELEASE_LOG_ERROR(Push, "WebClipCache::load failed to read %{public}@: %{public}@", path, error);
```

Comment:
Is it ok to log these paths, or could there be privacy concerns?

Reply:
Should be fine, it's a system path that doesn't contain PII.